[SPARK-13221] [SQL] Fixing GroupingSets when Aggregate Functions Containing GroupBy Columns

gatorsmile opened the pull request:
Using GroupingSets will generate a wrong result when Aggregate Functions containing GroupBy columns. 

This PR is to fix it. Since the code changes are very small. Maybe we also can merge it to 1.6

For example, the following query returns a wrong result:

``` scala
sql("select course, sum(earnings) as sum from courseSales group by course, earnings" +
     " grouping sets((), (course), (course, earnings))" +
     " order by course, sum").show()
```

Before the fix, the results are like

```
[null,null]
[Java,null]
[Java,20000.0]
[Java,30000.0]
[dotNET,null]
[dotNET,5000.0]
[dotNET,10000.0]
[dotNET,48000.0]
```

After the fix, the results become correct:

```
[null,113000.0]
[Java,20000.0]
[Java,30000.0]
[Java,50000.0]
[dotNET,5000.0]
[dotNET,10000.0]
[dotNET,48000.0]
[dotNET,63000.0]
```

UPDATE:  This PR also deprecated the external column: GROUPING__ID.

gatorsmile committed:
fixing grouping sets

SparkQA commented:
**[Test build #50855 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50855/consoleFull)** for PR 11100 at commit [`2f9eeb9`](https://github.com/apache/spark/commit/2f9eeb9d40e653119372a6d3b38106035f4fece9).

SparkQA commented:
**[Test build #50855 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50855/consoleFull)** for PR 11100 at commit [`2f9eeb9`](https://github.com/apache/spark/commit/2f9eeb9d40e653119372a6d3b38106035f4fece9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50855/
Test FAILed.

gatorsmile committed:
fix the issue when users manually specify grouping__id in the query.

SparkQA commented:
**[Test build #50869 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50869/consoleFull)** for PR 11100 at commit [`18f4130`](https://github.com/apache/spark/commit/18f413057b57efb3976dd04a0dfb723cdf59faea).

SparkQA commented:
**[Test build #50869 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50869/consoleFull)** for PR 11100 at commit [`18f4130`](https://github.com/apache/spark/commit/18f413057b57efb3976dd04a0dfb723cdf59faea).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50869/
Test FAILed.

gatorsmile commented:
Test failures are caused by flaky tests. 
retest this please.

gatorsmile commented:
retest this please

gatorsmile commented:
test this please

SparkQA commented:
**[Test build #50872 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50872/consoleFull)** for PR 11100 at commit [`18f4130`](https://github.com/apache/spark/commit/18f413057b57efb3976dd04a0dfb723cdf59faea).

SparkQA commented:
**[Test build #50872 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50872/consoleFull)** for PR 11100 at commit [`18f4130`](https://github.com/apache/spark/commit/18f413057b57efb3976dd04a0dfb723cdf59faea).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50872/
Test FAILed.

gatorsmile commented:
retest this please

SparkQA commented:
**[Test build #50892 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50892/consoleFull)** for PR 11100 at commit [`18f4130`](https://github.com/apache/spark/commit/18f413057b57efb3976dd04a0dfb723cdf59faea).

SparkQA commented:
**[Test build #50892 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50892/consoleFull)** for PR 11100 at commit [`18f4130`](https://github.com/apache/spark/commit/18f413057b57efb3976dd04a0dfb723cdf59faea).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50892/
Test FAILed.

gatorsmile commented:
retest this please

gatorsmile commented:
cc @yhuai @marmbrus @rxin  Could you check if this one is an appropriate fix? After checking the history, I found you are involved the discussion in the original fix of rollup and cube. : )

@liancheng This is blocking the JIRA https://issues.apache.org/jira/browse/SPARK-12720. Will submit a PR after this issue is addressed. Sorry for the delays. Thanks!

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -215,14 +215,31 @@ class Analyzer(
+        val aggExprs = g.aggregations.map(_.transform {
+          case u: UnresolvedAttribute if resolver(u.name, VirtualColumn.groupingIdName) => gid
+        }.asInstanceOf[NamedExpression])
+        g.copy(aggregations = aggExprs, groupByExprs = g.groupByExprs :+ gid)
```
This one has a potential bug. Will fix it soon.

SparkQA commented:
**[Test build #50907 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50907/consoleFull)** for PR 11100 at commit [`18f4130`](https://github.com/apache/spark/commit/18f413057b57efb3976dd04a0dfb723cdf59faea).

SparkQA commented:
**[Test build #50907 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50907/consoleFull)** for PR 11100 at commit [`18f4130`](https://github.com/apache/spark/commit/18f413057b57efb3976dd04a0dfb723cdf59faea).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50907/
Test FAILed.

gatorsmile committed:
fixed a bug.

SparkQA commented:
**[Test build #50909 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50909/consoleFull)** for PR 11100 at commit [`114e0eb`](https://github.com/apache/spark/commit/114e0eba1fbdf28e359f15bedacb7defa1998be6).

SparkQA commented:
**[Test build #50909 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50909/consoleFull)** for PR 11100 at commit [`114e0eb`](https://github.com/apache/spark/commit/114e0eba1fbdf28e359f15bedacb7defa1998be6).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50909/
Test FAILed.

gatorsmile commented:
retest this please

SparkQA commented:
**[Test build #50910 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50910/consoleFull)** for PR 11100 at commit [`114e0eb`](https://github.com/apache/spark/commit/114e0eba1fbdf28e359f15bedacb7defa1998be6).

SparkQA commented:
**[Test build #50910 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50910/consoleFull)** for PR 11100 at commit [`114e0eb`](https://github.com/apache/spark/commit/114e0eba1fbdf28e359f15bedacb7defa1998be6).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50910/
Test FAILed.

gatorsmile commented:
retest this please

SparkQA commented:
**[Test build #50925 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50925/consoleFull)** for PR 11100 at commit [`114e0eb`](https://github.com/apache/spark/commit/114e0eba1fbdf28e359f15bedacb7defa1998be6).

SparkQA commented:
**[Test build #50925 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50925/consoleFull)** for PR 11100 at commit [`114e0eb`](https://github.com/apache/spark/commit/114e0eba1fbdf28e359f15bedacb7defa1998be6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50925/
Test PASSed.

gatorsmile commented:
retest this please

SparkQA commented:
**[Test build #51080 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51080/consoleFull)** for PR 11100 at commit [`114e0eb`](https://github.com/apache/spark/commit/114e0eba1fbdf28e359f15bedacb7defa1998be6).

SparkQA commented:
**[Test build #51080 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51080/consoleFull)** for PR 11100 at commit [`114e0eb`](https://github.com/apache/spark/commit/114e0eba1fbdf28e359f15bedacb7defa1998be6).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51080/
Test FAILed.

gatorsmile committed:
fixed the duplicate test case name.

gatorsmile commented:
@davies @hvanhovell @aray I just realized you just reviewed a similar PR: https://github.com/apache/spark/pull/10677. Could you also review this one? I just checked the results are still wrong in the latest code. Thanks!

SparkQA commented:
**[Test build #51087 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51087/consoleFull)** for PR 11100 at commit [`524dfa0`](https://github.com/apache/spark/commit/524dfa072df691c4c218758b14d01817722577b5).

SparkQA commented:
**[Test build #51087 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51087/consoleFull)** for PR 11100 at commit [`524dfa0`](https://github.com/apache/spark/commit/524dfa072df691c4c218758b14d01817722577b5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51087/
Test FAILed.

aray commented:
LGTM

gatorsmile committed:
Merge remote-tracking branch 'upstream/master' into groupingSets

gatorsmile committed:
test case fix.

SparkQA commented:
**[Test build #51100 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51100/consoleFull)** for PR 11100 at commit [`e62c3d0`](https://github.com/apache/spark/commit/e62c3d0f908eb219798c958a6af731ce2750fbb8).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51100/
Test FAILed.

shaneknapp commented:
jenkins, test this please

SparkQA commented:
**[Test build #51111 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51111/consoleFull)** for PR 11100 at commit [`e62c3d0`](https://github.com/apache/spark/commit/e62c3d0f908eb219798c958a6af731ce2750fbb8).

SparkQA commented:
**[Test build #51111 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51111/consoleFull)** for PR 11100 at commit [`e62c3d0`](https://github.com/apache/spark/commit/e62c3d0f908eb219798c958a6af731ce2750fbb8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51111/
Test PASSed.

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -216,14 +216,36 @@ class Analyzer(
         GroupingSets(bitmasks(r), groupByExprs, child, aggregateExpressions)
-      case x: GroupingSets =>
+      // Ensure all the expressions have been resolved.
+      case g: GroupingSets if g.expressions.exists(!_.resolved) =>
```
g.expressions.forall(_.resolved)

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -216,14 +216,36 @@ class Analyzer(
+        else {
+          g
+        }
+      case x: GroupingSets =>
```
Add `if g.expressions.forall(_.resolved)` to make sure that all the expression are all resolved.

davies commented:
@gatorsmile Thanks for work on this.

The secret column name `GROUPING__ID` is only introduced by Hive, unfortunately the implementation is wrong, we don't follow that(corrected it). We can't be compatible with Hive anyway, so I'd like to not support it (it's OK for 2.0), we can have an error message to tell user to use `grouping_id()`.

For the other bug, that could be fixed by resolve all the expressions before GroupingSet (one line change).

davies commented:
@gatorsmile checked that your two tests could pass with these two tiny changes.

gatorsmile committed:
address comments.

gatorsmile commented:
Thank you! @davies @aray 

Yeah, my first fix is very similar to what you proposed above. Will remember what you said regarding `GROUPING__ID`. After the release of 2.0, I will try to deprecate it and issue an error message. 

BTW, just tried the code changes and it works well in my local environment. Updated the codes. Thanks!

SparkQA commented:
**[Test build #51169 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51169/consoleFull)** for PR 11100 at commit [`79c11de`](https://github.com/apache/spark/commit/79c11de8954e137e134d3a8645b6936cd625f38e).

SparkQA commented:
**[Test build #51169 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51169/consoleFull)** for PR 11100 at commit [`79c11de`](https://github.com/apache/spark/commit/79c11de8954e137e134d3a8645b6936cd625f38e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51169/
Test PASSed.

davies commented:
@gatorsmile I think the first commit should be enough. Since 2.0 is the best chance to deprecate GROUPING__ID, we should do that BEFORE release 2.0.

gatorsmile commented:
Sure, will revert the changes. Thank you!

gatorsmile committed:
address comments.

gatorsmile committed:
removed the test case using grouping__id

SparkQA commented:
**[Test build #51197 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51197/consoleFull)** for PR 11100 at commit [`ed518f9`](https://github.com/apache/spark/commit/ed518f946f83ff4d769ca3c095bfa25d406e4a3f).

SparkQA commented:
**[Test build #51197 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51197/consoleFull)** for PR 11100 at commit [`ed518f9`](https://github.com/apache/spark/commit/ed518f946f83ff4d769ca3c095bfa25d406e4a3f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51197/
Test FAILed.

gatorsmile committed:
removed GROUPING__ID from the test cases

SparkQA commented:
**[Test build #51229 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51229/consoleFull)** for PR 11100 at commit [`7631371`](https://github.com/apache/spark/commit/763137103ea5470a5fa9c14d02b4b0c9d1a32ed1).

SparkQA commented:
**[Test build #51229 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51229/consoleFull)** for PR 11100 at commit [`7631371`](https://github.com/apache/spark/commit/763137103ea5470a5fa9c14d02b4b0c9d1a32ed1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51229/
Test PASSed.

gatorsmile commented:
@davies Already deprecated `GROUPING__ID` in the latest commits and also clean all the test cases. 

Also tried to output a better error message when users manually specify `GROUPING__ID` in the query, but I hit an issue. To compare the column names, we need to pass `conf` or `resolver` to the function `checkAnalysis`. This will touch multiple files that call this function. I am not sure if it is worthy for this purpose. Let me know if you want me to do it. Thanks!
```
  def resolver: Resolver = {
    if (conf.caseSensitiveAnalysis) {
      caseSensitiveResolution
    } else {
      caseInsensitiveResolution
    }
  }
```

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -216,6 +216,8 @@ class Analyzer(
         GroupingSets(bitmasks(r), groupByExprs, child, aggregateExpressions)
+      // Ensure all the expressions have been resolved.
+      case g: GroupingSets if g.expressions.exists(!_.resolved) => g
       case x: GroupingSets =>
```
it's more clear if you move the `if` to next case

gatorsmile replied:
Ok, will do. Thanks!

davies commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala:
```diff
@@ -125,7 +125,7 @@ class HiveQuerySuite extends HiveComparisonTest with BeforeAndAfter {
 
   createQueryTest("SPARK-8976 Wrong Result for Rollup #1",
     """
-      SELECT count(*) AS cnt, key % 5,GROUPING__ID FROM src group by key%5 WITH ROLLUP
```
use `grouping_id()` instead

gatorsmile replied:
Will do. Thanks!

gatorsmile replied:
Hive does not have such a function.

```
SPARK-8976 Wrong Result for GroupingSet *** FAILED *** (316 milliseconds)
[info]   Failed to generate golden answer for query:
[info]   Error: FAILED: SemanticException [Error 10011]: Line 5:8 Invalid function 'grouping_id'
[info]   org.apache.spark.sql.execution.QueryExecutionException: FAILED: SemanticException [Error 10011]: Line 5:8 Invalid function 'grouping_id'
```

Hive

davies replied:
Since we already have the results, just copy them as golden files (or copy them into test cases).

gatorsmile replied:
Got it. Thanks!

davies commented:
@gatorsmile I think you can check `GROUPING__ID` in ResolveGroupingAnalytics, then raise an error

gatorsmile committed:
address comments.

gatorsmile committed:
removed unnecessary golden result files.

gatorsmile commented:
@davies As you suggested, the latest commit has the following changes:
- Replaced `GROUPING__ID` by `grouping_id()` in all the test cases and move the results into the test cases
- Moved all these test cases in `hive/execution/HiveQuerySuite.scala` to `hive/execution/SQLQuerySuite.scala`. 
- Detect the possible usage errors in the rule `ResolveGroupingAnalytics` and issue an error message if necessary

Thanks!

SparkQA commented:
**[Test build #51273 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51273/consoleFull)** for PR 11100 at commit [`1ca66ac`](https://github.com/apache/spark/commit/1ca66acc4d5cd962124942adcb2411d5c8566081).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51272/
Test FAILed.

hvanhovell commented:
retest this please

SparkQA commented:
**[Test build #51273 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51273/consoleFull)** for PR 11100 at commit [`1ca66ac`](https://github.com/apache/spark/commit/1ca66acc4d5cd962124942adcb2411d5c8566081).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51273/
Test PASSed.

SparkQA commented:
**[Test build #51275 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51275/consoleFull)** for PR 11100 at commit [`1ca66ac`](https://github.com/apache/spark/commit/1ca66acc4d5cd962124942adcb2411d5c8566081).

hvanhovell commented:
Whoops triggered build unnecessarily

hvanhovell commented:
@gatorsmile do we still need `VirtualColumn`? Other than that LGTM.

gatorsmile commented:
@hvanhovell Thank you for your reviews! 

Although we deprecate `GROUPING__ID`, the new `grouping_id()` still requires `VirtualColumn`. 

I like this change. Users are not allowed to select/query this hidden/secret column now.

SparkQA commented:
**[Test build #51275 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51275/consoleFull)** for PR 11100 at commit [`1ca66ac`](https://github.com/apache/spark/commit/1ca66acc4d5cd962124942adcb2411d5c8566081).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51275/
Test PASSed.

davies commented:
LGTM, we keep the `VirtualColumn` to show a better error message, merging this into master, thanks!
